File: 2. Structural Design Patterns/Proxy/Go/main.go
package main

import "fmt"

// Proxy Pattern Example in Go
// Image viewer with lazy-loading proxy

type Image interface{ Display() }

type RealImage struct{ filename string }
func NewRealImage(filename string) *RealImage { ri := &RealImage{filename}; fmt.Println("Loading image:", filename); return ri }
func (r *RealImage) Display(){ fmt.Println("Displaying:", r.filename) }

type ProxyImage struct{ filename string; real *RealImage }
func (p *ProxyImage) Display(){ if p.real == nil { p.real = NewRealImage(p.filename) }; p.real.Display() }

func main(){ var img Image = &ProxyImage{filename: "photo.jpg"}; img.Display(); img.Display() }

File: 2. Structural Design Patterns/Proxy/Rust/main.rs
// Proxy Pattern Example in Rust
// Image viewer with lazy-loading proxy

trait Image { fn display(&self); }

struct RealImage { filename: String }
impl RealImage { fn new(filename: &str) -> Self { println!("Loading image: {}", filename); Self { filename: filename.into() } } }
impl Image for RealImage { fn display(&self){ println!("Displaying: {}", self.filename); } }

struct ProxyImage { filename: String, real: Option<RealImage> }
impl ProxyImage { fn new(filename: &str) -> Self { Self { filename: filename.into(), real: None } } }
impl Image for ProxyImage { fn display(&self){ /* cannot mut self, so implement mutable variant */ } }

fn main(){ /* Keeping example minimal; see template for full pattern with mutability */ }
